src/Servers/Kestrel/Core/src/Internal/Http/Http1OutputProducer.cs

```diff
@@ -153,20 +229,33 @@ public void Abort(ConnectionAbortedException error)
             }
         }
 
-        public Task Write100ContinueAsync()
+        public ValueTask<FlushResult> Write100ContinueAsync()
         {
             return WriteAsync(_continueBytes.Span);
         }
 
-        private Task WriteAsync(
+        public void Complete(Exception exception = null)
+        {
+            // TODO What to do with exception.
```

Comment:
I would treat it as closely as possible to an application throwing an uncaught exception as possible. If the response hasn't started, that means a 500 response. Otherwise, log and and abort the underlying connection.

Reply:
So some combination of https://github.com/aspnet/AspNetCore/blob/20ee67b0cec738e99ccf4007876a729083804982/src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs#L585-L586 and making sure this logic isn't called twice?

Reply:
SetBadRequestState is used to set 4XX responses for invalid requests. I think callin

```C#
ReportApplicationError(ex);

if (HasResponseStarted)
{
    ApplicationAbort();
}
```

I would probably move ApplicationAbort() from HttpProtocol.FeatureCollection.cs to HttpProtocol.cs as part of this.

Reply:
Alright, I'll try that and see how it looks. @davidfowl I'll test this in a console app to see how visible the exception is too.

Reply:
@halter73 If someone called Complete() without an exception, we probably shouldn't treat this as the application error.

Reply:
Absolutely agreed. I was only talking about the exceptional case.

Reply:
Do not call the OnCompleted callbacks early, only ProduceEnd.

Reply:
I don't think we can call ProduceEnd as that writes to the response. Complete is sync and we can't flush/write t all. Also, I'd rather just let it be handled by the app func returning and handling it in ProcessRequests.